crates/bevy_render/src/render_resource/bind_group.rs

```diff
@@ -42,3 +53,255 @@ impl Deref for BindGroup {
+///     required.
+/// * `bind_group_data(DataType)`
+///     * The [`AsBindGroup`] type will be converted to some `DataType` using [`Into<DataType>`] and stored
+///     as [`AsBindGroup::Data`] as part of the [`AsBindGroup::as_bind_group`] call. This is useful if data needs to be stored alongside
+///     the generated bind group, such as a unique identifier for a material's bind group. The most common use case for this attribute
+///     is "shader pipeline specialization". See [`SpecializedRenderPipeline`](crate::render_resource::SpecializedRenderPipeline).
+///
+/// The previous `CoolMaterial` example illustrating "combining multiple field-level uniform attributes with the same binding index" can
+/// also be equivalently represented with a single struct-level uniform attribute:
+/// ```
+/// # use bevy_render::{color::Color, render_resource::{AsBindGroup, ShaderType}};
+/// #[derive(AsBindGroup)]
+/// #[uniform(0, CoolMaterialUniform)]
+/// struct CoolMaterial {
+///     color: Color,
+///     roughness: f32,
+/// }
+///
+/// #[derive(ShaderType)]
+/// struct CoolMaterialUniform {
+///     color: Color,
+///     roughness: f32,
+/// }
+///
+/// impl From<&CoolMaterial> for CoolMaterialUniform {
+///     fn from(material: &CoolMaterial) -> CoolMaterialUniform {
+///         CoolMaterialUniform {
+///             color: material.color,
+///             roughness: material.roughness,
+///         }
+///     }
+/// }
+/// ```
+///
+/// Setting `bind_group_data` looks like this:
+/// ```
+/// # use bevy_render::{color::Color, render_resource::AsBindGroup};
+/// #[derive(AsBindGroup)]
+/// #[bind_group_data(CoolMaterialKey)]
+/// struct CoolMaterial {
+///     #[uniform(0)]
+///     color: Color,
+///     is_shaded: bool,
+/// }
+///
+/// #[derive(Copy, Clone, Hash, Eq, PartialEq)]
+/// struct CoolMaterialKey {
+///     is_shaded: bool,
+/// }
+///
+/// impl From<&CoolMaterial> for CoolMaterialKey {
+///     fn from(material: &CoolMaterial) -> CoolMaterialKey {
+///         CoolMaterialKey {
+///             is_shaded: material.is_shaded,
+///         }
+///     }
+/// }
+/// ```
+pub trait AsBindGroup: Sized {
+    /// Data that will be stored alongside the "prepared" bind group.
+    type Data: Send + Sync;
+
+    /// Creates a bind group for `self` matching the layout defined in [`AsBindGroup::bind_group_layout`].
+    fn as_bind_group(
+        &self,
+        layout: &BindGroupLayout,
+        render_device: &RenderDevice,
+        images: &RenderAssets<Image>,
+        fallback_image: &FallbackImage,
+    ) -> Result<PreparedBindGroup<Self>, AsBindGroupError>;
+
+    /// Creates the bind group layout matching all bind groups returned by [`AsBindGroup::as_bind_group`]
+    fn bind_group_layout(render_device: &RenderDevice) -> BindGroupLayout;
+}
+
+/// An error that occurs during [`AsBindGroup::as_bind_group`] calls.
+pub enum AsBindGroupError {
```

Comment:
Should this be marked non_exhaustive? I would expect we'll expand this in the future, which is why we're using an enum with a single variant.

Reply:
Worth discussing. I haven't fully solidified my opinions on "non_exhaustive for public error types". If a new error condition is added and someone chose to handle AsBindGroupError exhaustively, maybe that _should_ be a breaking change.

Reply:
Why not let people choose to handle this exhaustively?

Reply:
The main argument is that it reduces breaking changes. In this case I think that's actually a negative; if we're introducing new error types I want downstream code to be forced to think about how they respond.

Reply:
Yeah I think non_exhaustive is more important for projects that don't want to do a "breaking release" for things like this. But all Bevy releases are breaking, so it matters less at this point in time. For now, I think we should leave it as is. We should come to a conclusion on our non_exhaustive policy for bevy errors on its own time.